llvm/include/llvm/Support/Progress.h

```diff
@@ -0,0 +1,256 @@
+//===-- Progress - Generic infrastructure to report progress ----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SUPPORT_PROGRESS_H
+#define LLVM_SUPPORT_PROGRESS_H
+
+extern "C" {
+#include <unistd.h>
+}
+
+#include <algorithm>
+#include <chrono>
+#include <optional>
+#include <set>
+#include <vector>
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/ManagedStatic.h"
+#include "llvm/Support/RWMutex.h"
+
+namespace llvm {
+
+class Progress;
+class TaskStack;
+
+/// Track advancement of a task
+///
+/// A task can have a name and optionally a set of maximum steps.
+class Task {
+protected:
+  TaskStack &Stack;
+  int64_t StepIndex = -1;
+  std::optional<int64_t> TotalSteps;
+
+private:
+  std::string TaskName;
+  std::string StepName;
+  bool SubtaskCreated = false;
+  bool Completed = false;
+  bool CurrentStepHasSingleSubtask = false;
```

Comment:
Is this very important? What is it used for?

From the code it's not immediately obvious what's it's used to accomplish, and it seems to me that dropping it would make the rest more clear to understand and with less special cases.

Maybe there's a way to implement the same functionality without treating this as special case?

Reply:
It is a bit subtle why we need this, but we need it.

If when you call `advance` you guarantee that there will be 0 or 1 subtasks, I can took into account the advancement of the single subtask in the advancement of the parent task. If you don't, I can't influence the parent task's advancement with the child task.

This is important because I expect that in VSCode we'll show a single top level progress bar.
Thanks to this feature, even if the top-level tasks has coarse granularity, maybe one single gigantic element, we can still show users advancement.

Reply:
So you're basically using this to make advancements of sub-tasks visible in the advancement of containing tasks. I see why this is something that can be desirable.

But at this point I'm gonna try to push for the other extreme: isn't there a way, without adding too much additional complexity, to enable this for all tasks and sub-tasks and not just for tasks with single sub-tasks?

Reply:
If the step of a task (i.e., `advance` invocation) can have an unknown number of subtasks you cannot tell how much progress you made in that step. Either 1) you assume there's always going to be at most one subtask, 2) you assume there's always more than one subtask, or 3) you ask the user to tell you if there's going to be more than one subtask.
1) does not work, it means that all of the codebase needs to guarantee that and with passes invoking other pass managers this is not going to happen any time and soon and hinders adoption.
2) means we can never propagate up progress.
3) is a compromise where the user actively opts-out of this if the current step is not `Progress.h` aware.

I mean, you could roll back advancement at every new subtask, but I've designed this to ensure we never ever ever rollback advancement, which I find frustrating.

Reply:
Eh you could have 4. you ask the user to tell you **exactly** how many sub-tasks there's gonna be. But I see how this is not practical in reality.

Let's keep this like it is then.

Reply:
> you ask the user to tell you exactly how many sub-tasks there's gonna be

If you could assume the collaboration of the user, you'd just ask him to create an intermediate task with the number of steps he expects. No need to introduce a new concept such as "multiple subtasks", just ask him to make an intermediate task.

The problem is that, for instance, we instrumented the pass manager, but not all of the passes. So certain passes end up having multiple subtasks (e.g., they run multiple subpipelines).

> Let's keep this like it is then.

OK.